File: apps/Mobile-App/split_the_bill/lib/utils/activity_navigation.dart
import '../models/models.dart';

const _billActivityTypes = {
  'BILL_CREATED',
  'BILL_UPDATED',
  'BILL_SETTLED',
  'BILL_UNSETTLED',
  'BILL_DELETED',
};

const _invitationActivityTypes = {
  'FRIEND_INVITATION_SENT',
  'FRIEND_INVITATION_ACCEPTED',
  'FRIEND_INVITATION_DECLINED',
};

String? activityRoute(ActivityEvent event) {
  if (_billActivityTypes.contains(event.type) && event.billId != null) {
    return '/bills/${event.billId}';
  }

  if (event.type == 'FRIEND_SETTLED' && event.friendshipId != null) {
    return '/friends/${event.friendshipId}';
  }

  if (_invitationActivityTypes.contains(event.type)) {
    return '/friends/invites';
  }

  return null;
}

File: apps/web/src/utils/activityNavigation.ts
import type { ActivityEvent } from '../api/types'

const billActivityTypes = new Set([
  'BILL_CREATED',
  'BILL_UPDATED',
  'BILL_SETTLED',
  'BILL_UNSETTLED',
])

const invitationActivityTypes = new Set([
  'FRIEND_INVITATION_SENT',
  'FRIEND_INVITATION_ACCEPTED',
  'FRIEND_INVITATION_DECLINED',
])

export function activityRoute(event: ActivityEvent): string | null {
  if (billActivityTypes.has(event.type) && event.billId) {
    return `/bills/${event.billId}`
  }

  if (event.type === 'FRIEND_SETTLED' && event.friendshipId) {
    return `/friends/${event.friendshipId}`
  }

  if (invitationActivityTypes.has(event.type)) {
    return '/friends?tab=invitations'
  }

  const groupId = (event as ActivityEvent & { groupId?: string | null }).groupId
  if (event.type.startsWith('GROUP_') && groupId && event.type !== 'GROUP_DELETED') {
    return `/groups/${groupId}`
  }

  return null
}

export function activityIsNavigable(event: ActivityEvent): boolean {
  return activityRoute(event) !== null
}
